mava/systems/tf/mad4pg/training.py

```diff
@@ -610,13 +608,27 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
                 clip_norm = True if self._max_gradient_norm is not None else False
 
                 policy_loss = losses.dpg(
-                    dpg_q_values,
-                    dpg_actions_comb,
+                    q_max=dpg_q_values,
+                    a_max=dpg_actions_comb,
                     tape=tape,
                     dqda_clipping=dqda_clipping,
                     clip_norm=clip_norm,
                 )
-                self.policy_losses[agent] = tf.reduce_mean(policy_loss, axis=0)
+                # Multiply by discounts to not train on padded data.
+                policy_mask = tf.reshape(agent_discount, policy_loss.shape) > 0.0
+                critic_mask = (
+                    tf.reshape(
+                        agent_discount[:, self._bootstrap_n :], critic_loss.shape
+                    )
+                    > 0.0
+                )
+                # TODO (dries): Is multiplication maybe better here? As assignment
```

Comment:
Not sure which assignment you are talking about?

Reply:
There are two ways of masking. On is multiplying with the mask and the other indexing by the mask. I am not sure which one is better as multiplying has some theoretical issues (but maybe not practically). Indexing is probably better. Just not sure if tf.functions works well with it.

Reply:
I agree with @KaleabTessera that we must be consistent. @DriesSmit what are the theoretical issues with multiplication?

Reply:
I updated the code with Claude's suggestion:) Thanks for pointing this out @arnupretorius  and @KaleabTessera  :raised_hands: